File: Week 1/id_295/Python/1_4/LeetCode_42_295.py
from typing import List


class Solution:
    def trap(self, height: List[int]) -> int:
        res, left, right, left_max, right_max = 0, 0, len(height) - 1, 0, 0
        while left < right:
            if height[left] < height[right]:
                if height[left] > left_max:
                    left_max = height[left]
                else:
                    res += left_max - height[left]
                left += 1
            else:
                if height[right] > right_max:
                    right_max = height[right]
                else:
                    res += right_max - height[right]
                right -= 1
        return res

File: Week 1/id_295/LeetCode_42_295.java
public class TrappingRainWater {

    class Solution {
        public int trap(int[] height) {
            int left = 0;
            int right = height.length - 1;
            int res = 0;
            int leftMax = 0;
            int rightMax = 0;

            while (left < right) {
                if (height[left] < height[right]) {
                    if (height[left] > leftMax) {
                        leftMax = height[left];
                    } else {
                        res += leftMax - height[left];
                    }
                    left++;
                } else {
                    if (height[right] > rightMax) {
                        rightMax = height[right];
                    } else {
                        res += rightMax - height[right];
                    }
                    right++;
                }
            }
            return res;
        }
    }
}
